README.md
# Introduction

This is a plugin for [Koha](http://koha-community.org) which enables the use of Capita Pay360 for online payments.

# Downloading


# Installing

## Enable the plugin system

To set up the Koha plugin system you must first make some changes to your install.

* Change `<enable_plugins>0<enable_plugins>` to `<enable_plugins>1</enable_plugins>` in your koha-conf.xml file
* Confirm that the path to `<pluginsdir>` exists, is correct, and is writable by the web server
* Add the pluginsdir to your apache PERL5LIB paths and koha-plack startup scripts PERL5LIB
* You will need to add the following to the apache config for your site:

```
   Alias /plugin/ "/var/lib/koha/kohadev/plugins/"
   # The stanza below is needed for Apache 2.4+
   <Directory /var/lib/koha/kohadev/plugins/>
         Options Indexes FollowSymLinks
         AllowOverride None
         Require all granted
         Options +ExecCGI
         AddHandler cgi-script .pl
    </Directory>
```

* Restart your webserver

Once set up is complete you will need to alter your UseKohaPlugins system preference. On the Tools page you will see the Tools Plugins and on the Reports page you will see the Reports Plugins.

## Dependancies

This plugin has some additional perl library dependancies which are not part of the standard Koha install and as such you will need to install them in your prefered manor:

* Data::GUID
* XML::Compile::WSDL11
* XML::Compile::SOAP
* XML::Compile::Transport::SOAPHTTP
* Time::Moment

`sudo apt install libxml-compile-wsdl11-perl libdata-guid-perl libtime-moment-perl` for example

## Download and install the plugin

The latest releases of this plugin can be obtained from the [release page](https://github.com/ptfs-europe/koha-plugin-capita-payments/releases) where you can download the relevant *.kpz file

# Setup

The available configuration options are all accessible from inside the staff client under 'Home › Tools › Plugins'



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v00.00.14] - 2022-07-15

## Fixed
- Added styling of payment results page

## [v00.00.13] - 2022-07-15

## Fixed
- Added a missing import

## [v00.00.12] - 2022-07-14

## Fixed
- Update plugin to work with Koha 21.11

## [v00.00.11] - 2020-06-19

### Fixed
- Update plugin to work with Koha 19.11 and 20.05 releases.

## [v00.00.10] - 2020-06-18

### Changed
- Updated configuration page styling to match modern practive
- Added the `interface` argument to the parameters passed in the call to Koha::Account::pay

## [v00.00.09] - 2019-07-05

### Added
- Ability to define a fundCode to be added globally to item payments

## [v00.00.08] - 2019-06-17

### Fixed
- Corrections to override of the client endpoint via the Pay360Portal preference.

## [v00.00.07] - 2019-06-14

### Fixed
- Allow override of the client endpoint via the Pay360Portal preference.

## [v00.00.06] - 2019-06-14

### Fixed
- Make Koha::Patrons->find call explicitly use userid.

## [v00.00.05] - 2019-04-12

### Fixed
- Respect renewal rules when renewal incrementing fines.

## [v00.00.04] - 2018-12-06

### Fixed
- Compilation error for resultset name typo.

## [v00.00.03] - 2018-12-06

### Added
- Add user email address to data in invokeRequest.

### Changed
- Remove `minlegth` from being appended to itemSummary/description fields.
- Use correct currency in opac end process.



Koha/Plugin/Com/PTFSEurope/CapitaPayments/callback.pl
#!/usr/bin/perl

# Copyright 2015 PTFS Europe
#
# This file is part of Koha.
#
# Koha is free software; you can redistribute it and/or modify it under the
# terms of the GNU General Public License as published by the Free Software
# Foundation; either version 3 of the License, or (at your option) any later
# version.
#
# Koha is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with Koha; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

use CGI qw( -utf8 );

use C4::Context;
use C4::Auth;
use Koha::Account;
use Koha::Account::Lines;
use Koha::Patrons;
use Koha::Plugin::Com::PTFSEurope::WPMPayments;

use XML::LibXML;
use Digest::MD5 qw(md5_hex);

my $paymentHandler = Koha::Plugin::Com::PTFSEurope::WPMPayments->new;

# Parse XML
binmode STDIN, ':encoding(UTF-8)';
my $xml_string;
while (<STDIN>) {
    $xml_string .= $_;
}
my $xml;
eval { $xml = XML::LibXML->load_xml( string => $xml_string ); };
warn "error: " . $@ if $@;

my $borrowernumber = $xml->findvalue('/wpmpaymentrequest/customerid');
my $transaction_id = $xml->findvalue('/wpmpaymentrequest/transactionreference');
my $success        = $xml->findvalue('/wpmpaymentrequest/transaction/success');

my $borrower = Koha::Patrons->find($borrowernumber);

if ( $success eq '1' ) {

    # Extract accountlines to pay
    my @accountline_ids = ();
    my $payments_nodes  = $xml->findnodes('/wpmpaymentrequest/payments');
    for my $payments_node ( $payments_nodes->get_nodelist ) {
        my $payment_nodes = $payments_node->findnodes('./payment');
        for my $payment_node ( $payment_nodes->get_nodelist ) {
            if ( $payment_node->findvalue('./@paid') ) {
                my $accountline = $payment_node->findvalue('./@payid');
                push @accountline_ids, $accountline;
            }
        }
    }

    my $totalpaid = $xml->findvalue('/wpmpaymentrequest/transaction/totalpaid');

    # Make Payment
    my $lines = Koha::Account::Lines->search(
        { accountlines_id => { 'in' => \@accountline_ids } } )->as_list;
    my $account = Koha::Account->new( { patron_id => $borrowernumber } );
    my $accountline_id = $account->pay(
        {
            amount     => $totalpaid,
            note       => 'WPM Payment',
            library_id => $borrower->branchcode,
            lines => $lines,    # Arrayref of Koha::Account::Line objects to pay
                                #account_type => $type,  # accounttype code
                 #offset_type  => $offset_type,    # offset type code
        }
    );

    # Link payment to wpm_transactions
    my $dbh   = C4::Context->dbh;
    my $table = $paymentHandler->get_qualified_table_name('wpm_transactions');
    my $sth   = $dbh->prepare(
        "UPDATE $table SET accountline_id = ? WHERE transaction_id = ?");
    $sth->execute( $accountline_id, $transaction_id );

    # Renew any items as required
    for my $account ( @{$lines} ) {

        # Renew if required
        if ( defined( $account->accountline->accounttype )
            && $account->accountline->accounttype eq "FU" )
        {
            if (
                CheckIfIssuedToPatron(
                    $account->accountline->borrowernumber->borrowernumber,
                    $account->accountline->itemnumber->biblionumber
                )
              )
            {
                my $datedue = AddRenewal(
                    $account->accountline->borrowernumber->borrowernumber,
                    $account->accountline->itemnumber->itemnumber
                );
                C4::Circulation::_FixOverduesOnReturn(
                    $account->accountline->borrowernumber->borrowernumber,
                    $account->accountline->itemnumber->itemnumber
                );
            }
        }
    }

    # Respond with OK
    my $response = new CGI;
    my $reply    = XML::LibXML::Document->new( '1.0', 'utf-8' );
    my $root     = $reply->createElement("wpmmessagevalidation");
    my $md5      = $xml->findvalue('/wpmpaymentrequest/@msgid');
    $root->setAttribute( 'msgid' => "$md5" );

    my $validation = $reply->createElement('validation');
    $validation->appendTextNode("1");
    $root->appendChild($validation);

    my $validationmessage = $reply->createElement('validationmessage');
    my $success           = XML::LibXML::CDATASection->new("Success");
    $validationmessage->appendChild($success);
    $root->appendChild($validationmessage);

    $reply->setDocumentElement($root);

    print CGI->header('text/xml');
    print $reply->toString();
}
else {
    # Update transaction status
    #
    # Respond OK

    my $reply = XML::LibXML::Document->new( '1.0', 'utf-8' );
    my $root  = $reply->createElement("wpmmessagevalidation");
    my $md5   = $xml->findvalue('/wpmpaymentrequest/@msgid');
    $root->setAttribute( 'msgid' => "$md5" );

    my $validation = $reply->createElement('validation');
    $validation->appendTextNode("1");
    $root->appendChild($validation);

    my $validationmessage = $reply->createElement('validationmessage');
    my $success           = XML::LibXML::CDATASection->new("Success");
    $validationmessage->appendChild($success);
    $root->appendChild($validationmessage);

    $reply->setDocumentElement($root);

    print CGI->header('text/xml');
    print $reply->toString();
}


Koha/Plugin/Com/PTFSEurope/CapitaPayments/scpService.pm
package Koha::Plugin::Com::PTFSEurope::CapitaPayments::scpService;
# Generated by SOAP::Lite (v1.27) for Perl -- soaplite.com
# Copyright (C) 2000-2006 Paul Kulchenko, Byrne Reese
# -- generated at [Tue Nov 20 11:57:40 2018]
# -- generated from https://sbsctest.e-paycapita.com/scp/scpws/scpSimpleClient.wsdl
my %methods = (
'scpSimpleInvoke' => {
    endpoint => 'https://sbsctest.e-paycapita.com/scp/scpws/',
    soapaction => '',
    namespace => 'http://www.capita-software-services.com/scp/simple',
    #    parameters => [
    #  SOAP::Data->new(name => 'scpSimpleInvokeRequest', type => 'scpSimpleInvokeRequest', attr => {}),
    #], # end parameters
  }, # end scpSimpleInvoke
'scpSimpleQuery' => {
    endpoint => 'https://sbsctest.e-paycapita.com:443/scp/scpws',
    soapaction => '',
    namespace => 'http://www.capita-software-services.com/scp/simple',
    parameters => [
      SOAP::Data->new(name => 'scpSimpleQueryRequest', type => 'scpbase:scpQueryRequest', attr => {}),
    ], # end parameters
  }, # end scpSimpleQuery
'scpVersion' => {
    endpoint => 'https://sbsctest.e-paycapita.com:443/scp/scpws',
    soapaction => '',
    namespace => 'http://www.capita-software-services.com/scp/simple',
    parameters => [
      SOAP::Data->new(name => 'scpVersionRequest', type => 'scpVersionRequest', attr => {}),
    ], # end parameters
  }, # end scpVersion
); # end my %methods

use SOAP::Lite;
use Exporter;
use Carp ();

use vars qw(@ISA $AUTOLOAD @EXPORT_OK %EXPORT_TAGS);
@ISA = qw(Exporter SOAP::Lite);
@EXPORT_OK = (keys %methods);
%EXPORT_TAGS = ('all' => [@EXPORT_OK]);

sub _call {
    my ($self, $method) = (shift, shift);
    my $name = UNIVERSAL::isa($method => 'SOAP::Data') ? $method->name : $method;
    my %method = %{$methods{$name}};
    $self->proxy($method{endpoint} || Carp::croak "No server address (proxy) specified")
        unless $self->proxy;
    my @templates = @{$method{parameters}};
    my @parameters = ();
    foreach my $param (@_) {
        warn "param: $param";
        if (@templates) {
            warn "templates";
            my $template = shift @templates;
            my ($prefix,$typename) = SOAP::Utils::splitqname($template->type);
            warn "typename:" . $typename;
            my $method = 'as_'.$typename;
            # TODO - if can('as_'.$typename) {...}
            my $result = $self->serializer->$method($param, $template->name, $template->type, $template->attr);
            warn "result: " . $result;
            push(@parameters, $template->value($result->[2]));
        }
        else {
            warn "passing param as is";
            push(@parameters, $param);
        }
    }
    $self->endpoint($method{endpoint})
       ->ns($method{namespace})
       ->on_action(sub{qq!"$method{soapaction}"!});
  $self->serializer->register_ns("http://www.capita-software-services.com/scp/simple","simple");
  #$self->serializer->register_ns("https://support.capita-software.co.uk/selfservice/?commonFoundation","sch1");
  #$self->serializer->register_ns("http://schemas.xmlsoap.org/wsdl/","wsdl");
  $self->serializer->register_ns("http://www.capita-software-services.com/scp/base","scpbase");
  #$self->serializer->register_ns("http://www.capita-software-services.com/portal-api","sch0");
  warn "calling SUPER::call($method => @parameters)";
    my $serialized = $self->serializer->serialize($parameters[0]);
    warn $serialized;
    my $som = $self->SUPER::call($method => @parameters);
    if ($self->want_som) {
        return $som;
    }
    warn "is SOAP::SOM" if UNIVERSAL::isa($som => 'SOAP::SOM');
    if ($som->fault) { die $som->fault->faultstring };
    use Data::Dumper;
    my $root = $som->root;
    warn Dumper($root);
    warn "SOAP::SOM->result: " .$som->result; 
    UNIVERSAL::isa($som => 'SOAP::SOM') ? wantarray ? $som->paramsall : $som->result : $som;
}

sub BEGIN {
    no strict 'refs';
    for my $method (qw(want_som)) {
        my $field = '_' . $method;
        *$method = sub {
            my $self = shift->new;
            @_ ? ($self->{$field} = shift, return $self) : return $self->{$field};
        }
    }
}
no strict 'refs';
for my $method (@EXPORT_OK) {
    my %method = %{$methods{$method}};
    *$method = sub {
        warn "called $method\n";
        my $self = UNIVERSAL::isa($_[0] => __PACKAGE__)
            ? ref $_[0]
                ? shift # OBJECT
                # CLASS, either get self or create new and assign to self
                : (shift->self || __PACKAGE__->self(__PACKAGE__->new))
            # function call, either get self or create new and assign to self
            : (__PACKAGE__->self || __PACKAGE__->self(__PACKAGE__->new));
        $self->_call($method, @_);
    }
}

sub AUTOLOAD {
    my $method = substr($AUTOLOAD, rindex($AUTOLOAD, '::') + 2);
    return if $method eq 'DESTROY' || $method eq 'want_som';
    die "Unrecognized method '$method'. List of available method(s): @EXPORT_OK\n";
}

1;


Koha/Plugin/Com/PTFSEurope/CapitaPayments.pm
use utf8;

package Koha::Plugin::Com::PTFSEurope::CapitaPayments;

=head1 NAME

Koha::Plugin::Com::PTFSEurope::CapitaPayments

=cut

use Modern::Perl;

use base qw(Koha::Plugins::Base);

use C4::Context;
use C4::Circulation qw( CheckIfIssuedToPatron CanBookBeRenewed AddRenewal );
use C4::Auth qw( checkauth get_template_and_user );

use Koha;
use Koha::Account;
use Koha::Account::Lines;
use Koha::Patrons;

use Mojo::Util qw(b64_decode);
use Digest::SHA qw(hmac_sha256_base64);
use Time::Moment;
use File::Basename;
use Data::GUID;
use Data::Dumper;

use XML::Compile::WSDL11;
use XML::Compile::SOAP11;
use XML::Compile::Transport::SOAPHTTP;

## Here we set our plugin version
our $VERSION = "00.00.14";
our $debug   = 0;

## Here is our metadata, some keys are required, some are optional
our $metadata = {
    name            => 'Capita Online Payments Plugin',
    author          => 'Martin Renvoize',
    date_authored   => '2018-06-13',
    date_updated    => "2022-07-15",
    minimum_version => '17.11.00.000',
    maximum_version => '21.11.09.000',
    version         => $VERSION,
    description     => 'This plugin implements online payments using '
      . 'Capita Educations payments platform.',
};

our $root = dirname(__FILE__);
our $wsdl =
  XML::Compile::WSDL11->new( $root . '/CapitaPayments/scpSimpleClient.wsdl' );
$wsdl->importDefinitions( $root . '/CapitaPayments/scpSimple.xsd' );
$wsdl->importDefinitions( $root . '/CapitaPayments/scpBaseTypes.xsd' );
$wsdl->importDefinitions( $root . '/CapitaPayments/commonPaymentTypes.xsd' );
$wsdl->importDefinitions( $root . '/CapitaPayments/commonFoundationTypes.xsd' );

sub new {
    my ( $class, $args ) = @_;

    ## We need to add our metadata here so our base class can access it
    $args->{'metadata'} = $metadata;
    $args->{'metadata'}->{'class'} = $class;

    ## Here, we call the 'new' method for our base class
    ## This runs some additional magic and checking
    ## and returns our actual $self
    my $self = $class->SUPER::new($args);

    return $self;
}

sub _version_check {
    my ( $self, $minversion ) = @_;

    $minversion =~ s/(.*\..*)\.(.*)\.(.*)/$1$2$3/;

    my $kohaversion = Koha::version();

    # remove the 3 last . to have a Perl number
    $kohaversion =~ s/(.*\..*)\.(.*)\.(.*)/$1$2$3/;

    return ( $kohaversion > $minversion );
}

sub opac_online_payment {
    my ( $self, $args ) = @_;

    return $self->retrieve_data('enable_opac_payments') eq 'Yes';
}

=head2 opac_online_payment_begin

  Initiate online payment process

=cut

sub opac_online_payment_begin {
    my ( $self, $args ) = @_;
    my $cgi    = $self->{'cgi'};
    my $schema = Koha::Database->new()->schema();

    my ( $userid, $cookie, $sessionID ) = C4::Auth::checkauth( $cgi, 0, undef, 'opac' );
    warn "Found userid: " . $userid . "\n" if $debug;

    # Generate unique transaction id
    my $transactionGUID = Data::GUID->new->as_string;

    # Get the borrower
    my $borrower = Koha::Patrons->find( { userid => $userid } );

    # Construct return URI
    my $return = C4::Context->preference('OPACBaseURL')
      . "/cgi-bin/koha/opac-account-pay-return.pl";
    my $returnURL = URI->new($return);
    $returnURL->query_form(
        {
            payment_method => scalar $cgi->param('payment_method'),
            txn            => $transactionGUID
        }
    );
    $returnURL = $returnURL->as_string;

    # Construct cancel URI
    my $cancelURL =
      C4::Context->preference('OPACBaseURL') . "/cgi-bin/koha/opac-account.pl";

    # Credentials
    my $Pay360SiteID   = $self->retrieve_data('Pay360SiteID');
    my $Pay360HMACID   = $self->retrieve_data('Pay360HMACID');
    my $Pay360PortalID = $self->retrieve_data('Pay360PortalID');
    my $requestGUID    = Data::GUID->new->as_string;
    my $timeStamp      = Time::Moment->now_utc->strftime('%Y%m%d%H%M%S');
    my $currentDigest  = $self->get_digest(
        'CapitaPortal', $Pay360PortalID, $requestGUID, $timeStamp,
        'Original',     $Pay360HMACID
    );
    my $Pay360FundCode = $self->retrieve_data('Pay360FundCode');

    # items
    my $sum_amountInMinorUnits = 0;
    my @accountline_ids        = $cgi->multi_param('accountline');
    my $accountlines           = $schema->resultset('Accountline')
      ->search( { accountlines_id => \@accountline_ids } );
    my $items;
    for my $accountline ( $accountlines->all ) {
        my $amount = sprintf "%.2f", $accountline->amountoutstanding;
        $amount                 = $amount * 100;
        $sum_amountInMinorUnits = $sum_amountInMinorUnits + $amount;

        my $item = {
            itemSummary => {
                description => ( $accountline->description )
                ? $accountline->description
                : "Description not available",
                amountInMinorUnits   => $amount,
                reference            => $accountline->accountlines_id,
                displayableReference => $accountline->accountlines_id,
            },
            quantity => 1,
            lineId   => $accountline->accountlines_id,
            (
                defined($Pay360FundCode)
                ? (
                    lgItemDetails => {
                        fundCode => $Pay360FundCode
                    }
                  )
                : ()
              )
        };
        push @{$items}, $item;
    }

    my $request = {
        credentials => {
            subject => {
                subjectType => 'CapitaPortal',
                identifier  => $Pay360PortalID,
                systemCode  => 'SCP'
            },
            requestIdentification => {
                uniqueReference => $requestGUID,
                timeStamp       => $timeStamp
            },
            signature => {
                algorithm => 'Original',
                hmacKeyID => $Pay360HMACID,
                digest    => $currentDigest
            },
        },
        requestType => 'payOnly',
        requestId   => $transactionGUID,
        routing     => {
            returnUrl => $returnURL,
            backUrl   => $cancelURL,
            siteId    => $Pay360SiteID,
            scpId     => $Pay360PortalID,
        },
        panEntryMethod => 'ECOM',
        billing        => {
            cardHolderDetails => {
                cardHolderName => $borrower->firstname . " "
                  . $borrower->surname,
                contact => {
                    email => $borrower->email
                }
            }
        },
        sale => {
            saleSummary => {
                description        => 'Library Payment',
                amountInMinorUnits => $sum_amountInMinorUnits,
            },
            items => {
                item => $items
            }
        }
    };

    my $portal         = $self->retrieve_data('Pay360Portal');
    my $transport_hook = sub {
        my ( $request, $trace, $transporter ) = @_;
        $request->uri($portal);
        $trace->{user_agent}->request($request);
    };

    my $scpSimpleInvoke =
      $portal
      ? $wsdl->compileClient(
        operation      => 'scpSimpleInvoke',
        transport_hook => $transport_hook
      )
      : $wsdl->compileClient( operation => 'scpSimpleInvoke' );

    my $response = $scpSimpleInvoke->($request);

    # Handle errors
    if (
        exists(
            $response->{'scpSimpleInvokeResponse'}->{'invokeResult'}
              ->{'errorDetails'}
        )
      )
    {
        warn Dumper( $response->{'scpSimpleInvokeResponse'}->{'invokeResult'}
              ->{'errorDetails'} );
        exit;    #FIXME: Return error page here
    }

    # Sucess
    elsif ( $response->{'scpSimpleInvokeResponse'} ) {

        # Store the transaction
        my $dbh   = C4::Context->dbh;
        my $table = $self->get_qualified_table_name('pay360_transactions');
        my $sth =
          $dbh->prepare(
"INSERT INTO $table (`transaction_guid`, `transaction_reference`, `transaction_state`) VALUES (?, ?, ?)"
          );
        $sth->execute(
            $transactionGUID,
            $response->{'scpSimpleInvokeResponse'}->{'scpReference'},
            $response->{'scpSimpleInvokeResponse'}->{'transactionState'}
        );

        print $cgi->redirect(
            $response->{'scpSimpleInvokeResponse'}->{'invokeResult'}
              ->{'redirectUrl'} );
        exit;
    }

    # No response
    else {
        warn "Uncaught scpSimpleInvoke: \n";
        warn Dumper($response);
    }
}

=head2 opac_online_payment_end

  Complete online payment process

=cut

sub opac_online_payment_end {
    my ( $self, $args ) = @_;
    my $cgi = $self->{'cgi'};

    my ( $template, $borrowernumber ) = get_template_and_user(
        {
            template_name   => $self->mbf_path('opac_online_payment_end.tt'),
            query           => $cgi,
            type            => 'opac',
            authnotrequired => 0,
            is_plugin       => 1,
        }
    );

    my $transactionGUID = $cgi->param('txn');

    # Credentials
    my $Pay360SiteID   = $self->retrieve_data('Pay360SiteID');
    my $Pay360HMACID   = $self->retrieve_data('Pay360HMACID');
    my $Pay360PortalID = $self->retrieve_data('Pay360PortalID');
    my $requestGUID    = Data::GUID->new->as_string;
    my $timeStamp      = Time::Moment->now_utc->strftime('%Y%m%d%H%M%S');
    my $currentDigest  = $self->get_digest(
        'CapitaPortal', $Pay360PortalID, $transactionGUID, $timeStamp,
        'Original',     $Pay360HMACID
    );

    # Fetch scpReference
    my $table = $self->get_qualified_table_name('pay360_transactions');
    my $dbh   = C4::Context->dbh;
    my $sth   = $dbh->prepare(
        "SELECT transaction_reference FROM $table WHERE transaction_guid = ?");
    $sth->execute($transactionGUID);
    my ($scpReference) = $sth->fetchrow_array();

    # Construct request
    my $request = {
        credentials => {
            subject => {
                subjectType => 'CapitaPortal',
                identifier  => $Pay360PortalID,
                systemCode  => 'SCP'
            },
            requestIdentification => {
                uniqueReference => $requestGUID,
                timeStamp       => $timeStamp
            },
            signature => {
                algorithm => 'Original',
                hmacKeyID => $Pay360HMACID,
                digest    => $currentDigest
            },
        },
        siteId       => $Pay360SiteID,
        scpReference => $scpReference
    };

    my $portal         = $self->retrieve_data('Pay360Portal');
    my $transport_hook = sub {
        my ( $request, $trace, $transporter ) = @_;
        $request->uri($portal);
        $trace->{user_agent}->request($request);
    };

    my $scpSimpleQuery =
      $portal
      ? $wsdl->compileClient(
        operation      => 'scpSimpleQuery',
        transport_hook => $transport_hook
      )
      : $wsdl->compileClient( operation => 'scpSimpleQuery' );

    my $response = $scpSimpleQuery->($request);

    # Handle errors
    if (
        exists(
            $response->{'scpSimpleQueryResponse'}->{'paymentResult'}
              ->{'errorDetails'}
        )
      )
    {
        use Data::Dumper;
        warn Dumper( $response->{'scpSimpleQueryResponse'}->{'paymentResult'}
              ->{'errorDetails'} );
    }

    # Sucess
    else {

        my $saleSummary =
          $response->{'scpSimpleQueryResponse'}->{'paymentResult'}
          ->{'paymentDetails'}->{'saleSummary'};
        my @accountline_ids =
          map { $_->{'lineId'} } @{ $saleSummary->{'items'}->{'itemSummary'} };

        # Record Payment
        my $borrower = Koha::Patrons->find($borrowernumber);
        my $lines    = Koha::Account::Lines->search(
            { accountlines_id => { 'in' => \@accountline_ids } } )->as_list;
        my $totalpaid = 0;
        for my $line ( @{$lines} ) {
            my $amount = sprintf "%.2f", $line->amountoutstanding;
            $amount    = $amount * 100;
            $totalpaid = $totalpaid + $amount;
        }
        $totalpaid = $totalpaid / 100;
        my $account = Koha::Account->new( { patron_id => $borrowernumber } );
        my $accountline_id = $account->pay(
            {
                amount     => $totalpaid,
                note       => 'Pay360 Payment',
                library_id => $borrower->branchcode,
                lines      => $lines,
                interface  => 'opac',
            }
        );

        # Link payment to pay360_transactions
        my $sth = $dbh->prepare(
            "UPDATE $table SET accountline_id = ? WHERE transaction_guid = ?");
        $sth->execute( $accountline_id, $transactionGUID );

        # Renew any items as required
        unless ( $self->_version_check('20.05.00') ) {
            for my $line ( @{$lines} ) {
                my $item =
                  Koha::Items->find( { itemnumber => $line->itemnumber } );

                # Renew if required
                if ( $self->_version_check('19.11.00') ) {
                    if (   $line->debit_type_code eq "OVERDUE"
                        && $line->status ne "UNRETURNED" )
                    {
                        if (
                            C4::Circulation::CheckIfIssuedToPatron(
                                $line->borrowernumber, $item->biblionumber
                            )
                          )
                        {
                            my ( $renew_ok, $error ) =
                              C4::Circulation::CanBookBeRenewed(
                                $line->borrowernumber, $line->itemnumber, 0 );
                            if ($renew_ok) {
                                C4::Circulation::AddRenewal(
                                    $line->borrowernumber, $line->itemnumber );
                            }
                        }
                    }
                }
                else {
                    if ( defined( $line->accounttype )
                        && $line->accounttype eq "FU" )
                    {
                        if (
                            C4::Circulation::CheckIfIssuedToPatron(
                                $line->borrowernumber, $item->biblionumber
                            )
                          )
                        {
                            my ( $can, $error ) =
                              C4::Circulation::CanBookBeRenewed(
                                $line->borrowernumber, $line->itemnumber, 0 );
                            if ($can) {

                                # Fix paid for fine before renewal to prevent 
                                # call to _CalculateAndUpdateFine if 
                                # CalculateFinesOnReturn is set.
                                C4::Circulation::_FixOverduesOnReturn(
                                    $line->borrowernumber, $line->itemnumber );

                                # Renew the item
                                my $datedue =
                                  C4::Circulation::AddRenewal(
                                    $line->borrowernumber, $line->itemnumber );
                            }
                        }
                    }
                }
            }
        }

        # Output result
        if ( defined($totalpaid) ) {
            $totalpaid = sprintf "%.2f", $totalpaid;
            $template->param(
                borrower      => scalar Koha::Patrons->find($borrowernumber),
                message       => 'valid_payment',
                message_value => $totalpaid
            );
        }
        else {
            $template->param(
                borrower => scalar Koha::Patrons->find($borrowernumber),
                message  => 'no_amount'
            );
        }

        print $cgi->header();
        print $template->output();
    }
}

=head2 configure

  Configuration routine
  
=cut

sub configure {
    my ( $self, $args ) = @_;
    my $cgi = $self->{'cgi'};

    unless ( $cgi->param('save') ) {
        my $template = $self->get_template( { file => 'configure.tt' } );

        ## Grab the values we already have for our settings, if any exist
        $template->param(
            enable_opac_payments =>
              $self->retrieve_data('enable_opac_payments'),
            Pay360SiteID   => $self->retrieve_data('Pay360SiteID'),
            Pay360HMAC     => $self->retrieve_data('Pay360HMAC'),
            Pay360Portal   => $self->retrieve_data('Pay360Portal'),
            Pay360PortalID => $self->retrieve_data('Pay360PortalID'),
            Pay360HMACID   => $self->retrieve_data('Pay360HMACID'),
            Pay360FundCode => $self->retrieve_data('Pay360FundCode'),
        );

        print $cgi->header();
        print $template->output();
    }
    else {
        $self->store_data(
            {
                enable_opac_payments => $cgi->param('enable_opac_payments'),
                Pay360SiteID         => $cgi->param('Pay360SiteID'),
                Pay360Portal         => $cgi->param('Pay360Portal'),
                Pay360PortalID       => $cgi->param('Pay360PortalID'),
                Pay360HMACID         => $cgi->param('Pay360HMACID'),
                Pay360HMAC           => $cgi->param('Pay360HMAC'),
                Pay360FundCode       => $cgi->param('Pay360FundCode'),
                last_configured_by   => C4::Context->userenv->{'number'},
            }
        );
        $self->go_home();
    }
}

## This is the 'install' method. Any database tables or other setup that should
## be done when the plugin if first installed should be executed in this method.
## The installation method should always return true if the installation succeeded
## or false if it failed.
sub install() {
    my ( $self, $args ) = @_;

    my $table = $self->get_qualified_table_name('pay360_transactions');

    return C4::Context->dbh->do( "
        CREATE TABLE IF NOT EXISTS $table (
            `transaction_id` INT( 11 ) NOT NULL AUTO_INCREMENT,
            `transaction_guid` varchar(36) NOT NULL,
            `transaction_reference` TINYTEXT,
            `transaction_state` TINYTEXT,
            `accountline_id` INT( 11 ),
            `updated` TIMESTAMP,
            PRIMARY KEY (`transaction_id`)
        ) ENGINE = INNODB;
    " );
}

=head2 get_digest

  Internal routine for generating the signature digest for messages

=cut

sub get_digest {
    my $self = shift;
    my (
        $subjectType, $identifier, $uniqueReference,
        $timestamp,   $algorithm,  $Pay360HMACID
    ) = @_;

    my @set = (
        $subjectType, $identifier, $uniqueReference,
        $timestamp,   $algorithm,  $Pay360HMACID
    );
    my $Pay360HMAC = $self->retrieve_data('Pay360HMAC');

    my $data   = join( '!', @set );
    my $key    = b64_decode($Pay360HMAC);
    my $digest = hmac_sha256_base64( $data, $key );
    while ( length($digest) % 4 ) {
        $digest .= '=';
    }

    if ($debug) {
        warn "HmacKey: " . $Pay360HMAC;
        warn "HmacKeyId: " . $Pay360HMACID;
        warn "Subject Type: " . $subjectType;
        warn "Identifier: " . $identifier;
        warn "Unique Reference: " . $uniqueReference;
        warn "Algorithm: " . $algorithm;
        warn "Timestamp: " . $timestamp;
        warn "String to hash: " . $data;
        warn "Digest: " . $digest;
    }

    return $digest;
}

## This is the 'upgrade' method. It will be triggered when a newer version of a
## plugin is installed over an existing older version of a plugin
#sub upgrade {
#    my ( $self, $args ) = @_;
#
#    my $dt = dt_from_string();
#    $self->store_data(
#        { last_upgraded => $dt->ymd('-') . ' ' . $dt->hms(':') } );
#
#    return 1;
#}

## This method will be run just before the plugin files are deleted
## when a plugin is uninstalled. It is good practice to clean up
## after ourselves!
#sub uninstall() {
#    my ( $self, $args ) = @_;
#
#    my $table = $self->get_qualified_table_name('mytable');
#
#    return C4::Context->dbh->do("DROP TABLE $table");
#}

1;
